src/ucp/proto/proto_perf.c

```diff
@@ -53,35 +54,51 @@ struct ucp_proto_perf {
     ucs_list_for_each((_seg), &(_perf)->segments, list)
 
 static const char *ucp_proto_perf_factor_names[] = {
-    [UCP_PROTO_PERF_FACTOR_LOCAL_CPU]  = "cpu",
-    [UCP_PROTO_PERF_FACTOR_REMOTE_CPU] = "rem_cpu",
-    [UCP_PROTO_PERF_FACTOR_LOCAL_TL]   = "tl",
-    [UCP_PROTO_PERF_FACTOR_REMOTE_TL]  = "rem_tl",
-    [UCP_PROTO_PERF_FACTOR_LATENCY]    = "lat",
-    [UCP_PROTO_PERF_FACTOR_SINGLE]     = "sngl",
-    [UCP_PROTO_PERF_FACTOR_MULTI]      = "mult",
-    [UCP_PROTO_PERF_FACTOR_LAST]       = NULL
+    [UCP_PROTO_PERF_FACTOR_LOCAL_CPU]     = "cpu",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_CPU]    = "cpu-remote",
+    [UCP_PROTO_PERF_FACTOR_LOCAL_TL]      = "tl",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_TL]     = "tl-remote",
+    [UCP_PROTO_PERF_FACTOR_LOCAL_MTCOPY]  = "mtcopy",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_MTCOPY] = "mtcopy-remote",
+    [UCP_PROTO_PERF_FACTOR_LATENCY]       = "lat",
+    [UCP_PROTO_PERF_FACTOR_LAST]          = NULL
 };
 
 static void ucp_proto_perf_check(const ucp_proto_perf_t *perf)
 {
 #if ENABLE_ASSERT
+    ucs_string_buffer_t funcb = UCS_STRING_BUFFER_INITIALIZER;
+    const char *reason        = NULL;
     const ucp_proto_perf_segment_t *seg;
     size_t min_start;
 
     min_start = 0;
     ucp_proto_perf_segment_foreach(seg, perf) {
-        ucs_assertv((seg->start >= min_start) && (seg->start <= seg->end),
-                    "perf=%p seg->start=%zu seg->end=%zu min_start=%zu", perf,
-                    seg->start, seg->end, min_start);
-        if (seg->end == SIZE_MAX) {
-            ucs_assertv(ucs_list_is_last(&perf->segments, &seg->list),
-                        "perf=%p seg->start=%zu seg->end=%zu", perf, seg->start,
-                        seg->end);
-        } else {
+        if (seg->start < min_start) {
+            reason = "seg->start < min_start";
+            break;
+        }
+        if (seg->start > seg->end) {
+            reason = "seg->start > seg->end";
+            break;
+        }
+        if (seg->end < SIZE_MAX) {
             min_start = seg->end + 1;
+        } else if (!ucs_list_is_last(&perf->segments, &seg->list)) {
+            reason = "!ucs_list_is_last(&perf->segments, &seg->list)";
+            break;
         }
     }
+
+    if (reason != NULL) {
+        ucp_proto_perf_str(perf, &funcb);
+        ucs_debug("perf=%p name=%s %s", perf, perf->name,
+                  ucs_string_buffer_cstr(&funcb));
+        ucs_fatal("%s perf=%p seg->start=%zu seg->end=%zu min_start=%zu",
+                  reason, perf, seg->start, seg->end, min_start);
```

Comment:
maybe print everythin in `fatal`?

Reply:
can you try using \n in the message if need to split to multiple lines?

Reply:
the goal was just code simplification, but it is up to you

Reply:
Done